marxs/optics/multiLayerMirror.py

```diff
@@ -0,0 +1,104 @@
+import numpy as np
+from astropy.io import ascii
+
+from .base import FlatOpticalElement
+from ..math.pluecker import *
+
+
+class MultiLayerMirror(FlatOpticalElement):
+    '''Multilayer mirror with varying layer thickness along one axis
+    
+    The distance between layers changes along the local y axis.
+    All reflectivity data is assumed to be for a single, desired angle. There
+    is currently no way to enter varying reflection that depends on the angle
+    of incidence.
+    Provide reflectivity data in a file with columns:
+    	'X(mm)' - position along the "changing" axis
+    	'Peak lambda' - wavelength with maximum reflection at a given position
+    	'Peak' - maximum reflection at a given position
+    	'FWHM(nm)' - full width half max, measure of width of reflection Gaussian peaks
+    
+    Parameters
+    ----------
+    fileCode: string
+    	serial number of mirror, used to open reflection data file
+    '''
+    def __init__(self, fileCode, **kwargs):
+        self.fileName = './marxs/optics/docs/' + fileCode + '.txt'
+        if ('zoom' not in kwargs):
+        	kwargs['zoom'] = np.array([1, 24.5, 12])   # in mm
+        super(MultiLayerMirror, self).__init__(**kwargs)
+
+    def process_photons(self, photons):
+        # eliminate photons that do not reach the mirror
+        doesIntersect, intersection, local = self.intersect(photons['dir'], photons['pos'])
+        photons['probability'] *= doesIntersect
+        
+        # split polarization into parallel and other components
+        beam_dir = (photons['dir'][:,0:3]).copy()
+        for i in range(0, len(beam_dir)):
+        	beam_dir[i] /= np.linalg.norm(beam_dir[i])
```

Comment:
`beam_dir /= np.linalg.norm(beam_dir, axis=1)[:, np.newaxis]` can do this on the whole array at once. Google "Numpy fancy indexing" for an explanation why this works.
For a small number of photons your version is better, but for a large number of photons we should avoid the `for` loop because they are very slow in an interpreted language like python. Instead, we for loop is implicit in the array syntax (`np.linalg.norm(xxx, axis=1)` means: Loop of axis 0 and calculate norm of axis 1 for each subarray) and the numpy stuff is implemented directly in C which is much faster.